river/anomaly/zscore.py

```diff
@@ -0,0 +1,67 @@
+import math
+
+from river.stats.var import RollingVar
+
+from .base import AnomalyDetector
+
+
+class Zscore(AnomalyDetector):
+    """
+    Anomaly detection based on zscore. Data is assumed to have gaussian distribution
+
+    Examples
+    --------
+    import numppy as np
+    from river.anomaly.zscore import Zscore
+
+    #warmup anomaly detector
+    zs = Zscore(100)
+    mean = 10.0
+    sd = 2.0
+
+    for _ in range(110):
+        v = np.random.normal(mean, sd)
+        zs.learn_one(v)
+
+    #get score
+    v = mean + 4 * sd
+    sc = zs.score_one(v)
+    print(sc)
+
+    """
+
+    def __init__(self, window_size: int, zs_threshold: float = 3.0):
+        """
+
+        Parameters
+        ----------
+        window_size : size of rolling window
+        zs_threshold : zscore threshold
+        """
+        self.var = RollingVar(window_size)
+        self.zs_threshold = zs_threshold
+
+    def learn_one(self, x: float):
```

Comment:
i will change float to a dict. I used float because this algorithm is for the univariate data only. For multivariate data, the algorithms are different e.g Isolation Forest

Reply:
> I used float because this algorithm is for the univariate data only.

I have a soft conviction that we should support multivariate data by default, and not implement algorithms that only support univariate data. In this case, I suggest that you extend your algorithm to support multivariate data by multiplying the z-scores with each other.

Reply:
People have tried average, median or max to aggregate scores for muti variate data. That kind of pseudo multi variate Anomaly Detection(AD) works only if the co variance is insignificant. if not,  users will have to preprocess the data to do  dimension reduction and remove co variance through PCA or AutoEncoder.

If Zscore is not acceptable for univariate data only, you can reject the pull request. I will look into online learning for Isolation Forest, which is the most popular algorithm for multi variate anomaly detection. it may be lot of work. I have batch implementation in Spark. I have to figure out how to extend the implementation  for online.

BTW, in AD  it's common to use different algorithms as mentioned in my discussion early on. For example, Zscore will fail if the data is auto regressive. That's why  I proposed 6 different algorithms for AD  for  3 different type scenarios in the initial discussion. it's  the reason I implemented extensive EDA  in my repo that enables users to discover the kind of data they have.

Reply:
> If Zscore is not acceptable for univariate data only, you can reject the pull request.

Sadly I think we're going to have to reject the pull request. My opinion is that this method is too trivial, and that by default we should handle multivariate data. That doesn't all the work you put in doesn't matter though! You've learnt a lot and we've started a discussion.

> I will look into online learning for Isolation Forest, which is the most popular algorithm for multi variate anomaly detection. it may be lot of work. I have batch implementation in Spark. I have to figure out how to extend the implementation for online.

We already have `HalfSpaceTrees`, which are basically the online equivalent of an isolation forest. There's probably some improvements to bring to the table though.

Reply:
Fair enough. I had proposed 6 algorithms in discussion, 2 each for univariate, univariate auto regressive and multivariate. Your response was in agreement . If you had raised issues at that time, I didn't have to waste time working on this pull request. That's kind of puzzling to me.

Reply:
You're right, I have changed my mind. This simply shows I'm unsure of the way to proceed. That's the thing with open source: there are no project managers so strong ownership is necessary. It's true that this method brings some value, what I'm wondering is if there isn't a method based on Gaussian distributions which works for multi variate data. If there isn't, then so be it. And if we implement this method, then we need to expand the API by introducing a new base class and documenting the difference between univariate and multi variate methods.

Reply:
Since you are interested in multi variate AD only, there are several algorithms besides Isolation Forest, namely nearest neighbor, local outlier factor, clustering, auto encoder. Let me know if you are interested in any of them, or if there is some other area in river where contribution is needed. However I will spend time working on something only after you have thought through it and firmly committed to it.